index.js

```diff
@@ -763,25 +763,31 @@ function Runtime() {
                         return null;
                     }
 
-                    if ("- methodSignatureForSelector:" in target) {
-                        const s = target.methodSignatureForSelector_(sel);
-                        if (s === null)
-                            return null;
-                        const numArgs = s.numberOfArguments().valueOf();
-                        const frameSize = numArgs * pointerSize;
-                        let types = s.methodReturnType() + frameSize;
-                        for (let i = 0; i !== numArgs; i++) {
-                            const frameOffset = (i * pointerSize);
-                            types += s.getArgumentTypeAtIndex_(i) + frameOffset;
-                        }
-                        method = {
-                            sel: sel,
-                            types: types,
-                            wrapper: null
-                        };
-                    } else {
+                    const methodHandle = api.class_getInstanceMethod(api.object_getClass(target.handle), sel);
+                    if (methodHandle.isNull()) {
                         return null;
                     }
+                    let types = Memory.readUtf8String(api.method_getTypeEncoding(methodHandle));
+                    if (types === null || types === '') {
+                        types = stealTypesFromProtocols(target, fullName);
+                        if (types === null) {
+                            types = stealTypesFromProtocols(self, fullName);
+                        }
+                        if (types === null) {
```

Comment:
making types up may be unsafe, if the caller expects different types it may fail in ways which could be pretty difficult to spot (like for example, expecting an NSString as a return value the caller will do `retval.toString()` which in this case isn't failing but it will end up with a stringified pointer instead)